consensus/miner_signalled_consensus.md

```diff
@@ -0,0 +1,200 @@
+same height.  As a result, the user experience would be worse:
+- By taking the decision globally, on bootstrap a node may initially follow a
+  (valid) chain fork distinct from the best valid one hence take the wrong
+  decision.
+- As the best valid chain is not necessarily the longest one, by taking the
+  decision globally the node may anyway taint the storage with blocks with
+  eventually wrong consensus protocol.
+
+### Value of the info field not necessarily part of new consensus protocol
+
+The miner signalling is orthogonal to the validation rules of the new
+consensus protocol for the sake of simplicity of the implementations.
+
+This is not detrimental as the new consensus protocol may include
+arbitrary constraints.
+
+## Backwards Compatibility
+
+The signalling using the block header field `info` is backward compatible, as
+the meaning of the value of such field is not under consensus.
+
+The activation of the new consensus protocol based on miner signalling is meant
+to be a temporary measure during the time interval when the new protocol may be
+activated on the network while the node is running. Once the network reasonably
+settles on whether to activate the new consensus protocol, the user:
+- Is meant to enforce such decision on the eventual node restart either by
+  configuration or by upgrading the node to a new version.
+- Shall be recommended to discard the persisted storage on the eventual node
+  restart - as distinct persisted chain forks may have distinct consensus
+  protocols at the same height.
+
+In order to enable non-mutually authorized channel transactions to be
+included on chain with mutually-authorized-agreed-offline payload,
+channel transactions valid under the current consensus protocol must
+be valid also under the proposed new consensus protocol.
+
+## (possible) Future extensions
+
+### Simultaneous consensus proposals
+
+The usage of the block header field `info` offers room for optimization for
+accommodating simultaneous hard-forking or soft-forking consensus proposals, by
+interpreting such `info` field as a bitmask rather than a value.
+
+### Soft-forking consensus proposals
+
+This proposal may be extended to soft-forking consensus proposals - as opposed
+to only hard-forking ones - by relaxing the strict monotonicity of consensus
+versions.
+
+Such extension should preferably be accompanied by a working implementation.
+
+## Implementation
+
+Any implementation needs to cater for the performance of the aggregation of the
+signalling e.g. slow multiple lookups on the storage.
+
+Any implementation needs to consider that the storage may contain blocks with
+the old and new consensus protocols at heights at which the new consensus
+protocol should have been applied. This is because the node should be aware of
+multiple chain forks and peers.
+
+Any implementation needs to be resilient to node restarts with
+accidental distinct miner signalling parameters in the configuration.
```

Comment:
Since any hard fork will mean a new "release"/version of the node the _only_ configuration is whether to signal or not. This does not affect the history in any way. I.e. there is nothing that needs to be re-computed when you restart the node. The (partial) result of the vote could then be tallied in every block (not under consensus, but in the local database) as the node goes along - this means there is no need for an additional worker process and there is no need to have a time window after the vote period ends. 

I think this will simplify things considerably, but maybe I am missing something...

Reply:
The config should include more properties than "signal/not signal". There may be concurrent forks going on and we should make it possible to choose the signal value, etc.

If we wanted to keep the partial result in a local persistent db, then the result would apply just to one particular config. If we decide to restart the node, then the partial result might not be applicable anymore if we the config was changed. So we would need to recompute the result again and store it.

Reply:
> there is no need for an additional worker process and there is no need to have a time window after the vote period ends.

The time window after the fork signalling interval ends is there mainly because we want the chain between HS and HE to reasonably settle (so there is less likelihood of forks in the signalling interval and so the fork signalling result at H2 shouldn't change)

Reply:
I don't think the signalling outcome should depend on the configuration... Either you decide to follow the outcome of the "poll" and then it is a feature that all nodes behave the same (since the poll is on chain they will) - if you don't want to follow the outcome you better run an older/other version of the node? **We** don't want the network to split, so let's not give them the tools for this ;-)

Reply:
The configuration values would have to be hardcoded in the node then, and the only configuration option would be "follow signalling outcome/stay with the current protocol version".

I still prefer having the fork signalling configuration in the configuration file though. This is more flexible and it would allow the community to do protocol upgrade themselves (which may not be the best thing from our point of view). My understanding of fork signalling is that we just provide means of doing the protocol upgrades, we suggest some features and let the community decide. Maybe some other party can suggest new features as well, but they still will be able to use the fork signalling mechanism since it will be configurable.

This might be something we need to discuss further and more people should be involved...